cmd/mrregister.cpp

```diff
@@ -164,72 +200,138 @@ void run ()
-  } else if (im1_image.ndim() == 3) {
-    do_reorientation = false;
-  } else if (im1_image.ndim() == 4) {
-    if (im1_image.size(3) != im2_image.size(3))
-      throw Exception ("input images do not have the same number of volumes in the 4th dimension");
-    image_lmax = Math::SH::LforN (im1_image.size(3));
-    value_type val = (std::sqrt (float (1 + 8 * im2_image.size(3))) - 3.0) / 4.0;
-    if (!(val - (int)val) && do_reorientation && im2_image.size(3) > 1) {
-      CONSOLE ("SH series detected, performing FOD registration");
-      do_reorientation = true;
-      if (!directions_cartesian.cols())
-        directions_cartesian = Math::Sphere::spherical2cartesian (DWI::Directions::electrostatic_repulsion_60()).transpose();
-    } else {
-      do_reorientation = false;
-      if (directions_cartesian.cols())
-        WARN ("-directions option ignored since no FOD reorientation is being performed");
+  // check header transformations for equality
+  Eigen::MatrixXd trafo = MR::Transform(input1[0]).scanner2voxel.linear();
+  for (size_t i=1; i<n_images; i++) {
+    if (!trafo.isApprox(MR::Transform(input1[i]).scanner2voxel.linear(),1e-5))
+      WARN ("Multi contrast image has different header transformation from first image. Ignoring transformation of " + str(input1[i].name()));
+  }
+  trafo = MR::Transform(input2[0]).scanner2voxel.linear();
+  for (size_t i=1; i<n_images; i++) {
+    if (!trafo.isApprox(MR::Transform(input2[i]).scanner2voxel.linear(),1e-5))
+      WARN ("Multi contrast image has different header transformation from first image. Ignoring transformation of " + str(input2[i].name()));
+  }
+
+  // multi contrast settings
+  vector<Registration::MultiContrastSetting> mc_params (n_images);
+  for (auto& mc : mc_params) {
+    mc.do_reorientation = do_reorientation;
+  }
+  // set parameters for each contrast
+  for (size_t i=0; i<n_images; i++) {
+    // compare input1 and input2 for consistency
+    if (i>0) check_dimensions (input1[i], input1[i-1], 0, 3);
+    if (i>0) check_dimensions (input2[i], input2[i-1], 0, 3);
+    if ((input1[i].ndim() != 3) and (input1[i].ndim() != 4))
+      throw Exception ("image dimensionality other than 3 or 4 are not supported. image " +
+        str(input1[i].name()) + " is " + str(input1[i].ndim()) + " dimensional");
+
+    const size_t nvols1 = input1[i].ndim() == 3 ? 1 : input1[i].size(3);
+    const size_t nvols2 = input2[i].ndim() == 3 ? 1 : input2[i].size(3);
+    if (nvols1 != nvols2)
+      throw Exception ("input images do not have the same number of volumes: " + str(input2[i].name()) + " and " + str(input1[i].name()));
+
+    // set do_reorientation and image_lmax
+    if (nvols1 == 1) { // 3D or one volume
+      mc_params[i].do_reorientation = false;
+      mc_params[i].image_lmax = 0;
+      CONSOLE ("3D input pair "+input1[i].name()+", "+input2[i].name());
+    } else { // more than one volume
+      value_type val = (std::sqrt (float (1 + 8 * nvols1)) - 3.0) / 4.0;
```

Comment:
Similar to previous: an `SH::LforN()` followed by a modulo would make more sense when reading the code